pkg/build/generator/generator.go

```diff
@@ -93,6 +99,23 @@ func (g *BuildGenerator) Instantiate(ctx kapi.Context, request *buildapi.BuildRe
 	if err != nil {
 		return nil, err
 	}
+
+	// Fetch the secrets associated with the BuilderServiceAccount.
+	// These secrets will be used for push/pull authentication
+	// FIXME: Add support for SourceSecret
+	if sa, err := g.ServiceAccounts.ServiceAccounts(bc.Namespace).Get(bootstrappolicy.BuilderServiceAccountName); err != nil {
+		glog.V(2).Infof("Unable to fetch builder service account for %s: %v", bc.Namespace, err)
+	} else {
+		for _, ref := range sa.Secrets {
+			secret, err := g.Secrets.Secrets(bc.Namespace).Get(ref.Name)
+			if err != nil {
+				glog.V(5).Infof("Unable to fetch builder secret %s/%s: %v", bc.Namespace, ref.Name, err)
+				continue
+			}
+			g.builderSecrets = append(g.builderSecrets, *secret)
```

Comment:
`builderSecrets` is used here: https://github.com/openshift/origin/pull/2475/files#diff-82f54c5f4bf72a64f869a69e557df8cbR160 to locate a secret to use for an image pull (that's the only reason to make a `DockerKeyring`.  Given that intent, it seems reasonable to avoid using any random secret that this ServiceAccount can mount and instead limit the set of secrets to those that have been select to perform pulls.